datasnap-ios-sdk/DSIOGimbalClient.m

```diff
@@ -5,81 +5,91 @@
 //  Created by Alyssa McIntyre on 6/1/16.
 //  Copyright © 2016 Datasnapio. All rights reserved.
 //
-
+#import "BaseEvent.h"
+#import "Beacon.h"
+#import "Campaign.h"
+#import "Communication.h"
 #import "DSIOGimbalClient.h"
-static DSIOClient* sharedInstance = nil;
+#import "DataSnap.h"
+#import "Device.h"
+#import "DeviceInfo.h"
+#import "EventEntity.h"
+#import "Location.h"
+#import "Place.h"
+#import "User.h"
 @implementation DSIOGimbalClient
+
+- (void)startGimbal
+{
+    [Gimbal setAPIKey:self.gimbalApiKey options:nil];
+    [Gimbal start];
+    UIUserNotificationType types = UIUserNotificationTypeBadge | UIUserNotificationTypeSound | UIUserNotificationTypeAlert;
+    UIUserNotificationSettings* mySettings =
+        [UIUserNotificationSettings settingsForTypes:types categories:nil];
+    [[UIApplication sharedApplication] registerUserNotificationSettings:mySettings];
+    [[UIApplication sharedApplication] registerForRemoteNotifications];
+    self.beaconManager = [GMBLBeaconManager new];
+    self.beaconManager.delegate = self;
+    [self.beaconManager startListening];
+    self.communicationManager = [GMBLCommunicationManager new];
+    self.communicationManager.delegate = self;
+    [GMBLPlaceManager startMonitoring];
+    [GMBLCommunicationManager startReceivingCommunications];
+    DSIOBaseClient* baseClient = [[DSIOBaseClient alloc] init];
+    [baseClient dsioBaseClient];
+}
 - (void)beaconManager:(GMBLBeaconManager*)manager didReceiveBeaconSighting:(GMBLBeaconSighting*)sighting
 {
-    NSDateFormatter* dateFormatter = [[NSDateFormatter alloc] init];
-    [dateFormatter setDateFormat:@"yyyy-MM-dd'T'HH:mm:ssZZZZZ"];
-    NSDictionary* beaconData = @{ @"event_type" : @"beacon_sighting",
-        @"beacon" : @{ @"identifier" : sighting.beacon.identifier,
-            @"rssi" : [NSString stringWithFormat:@"%ld", (long)sighting.RSSI] },
-        @"datasnap" : @{ @"created" : [NSString stringWithFormat:@"%@", [dateFormatter stringFromDate:sighting.date]] },
-        @"user" : @{
-            @"id" : @{
-                @"global_distinct_id" : self.global_distinct_id,
-                @"mobile_device_ios_idfa" : self.mobile_device_ios_idfa
-            }
-        }
-    };
+    NSString* eventType = @"beacon_sighting";
+    Beacon* beacon = [[Beacon alloc] init];
+    beacon.identifier = sighting.beacon.identifier;
+    beacon.rssi = sighting.beacon.uuid;
+    beacon.batteryLevel = [NSString stringWithFormat:@"%ld", (long)sighting.beacon.batteryLevel];
+    beacon.name = sighting.beacon.name;
+    beacon.bleVendorId = @"Gimbal";
+    self.user.identifier.globalDistinctId = self.global_distinct_id;
+    self.user.identifier.mobileDeviceIosIdfa = self.mobile_device_ios_idfa;
+    BeaconEvent* event = [[BeaconEvent alloc] init];
```

Comment:
There are instances of creating Beacon & Communications events throughout and setting properties on them... Even if there are different properties, could you create an EventManager or something to build and track them? Something like buildEventOfType:(NSString*)eventType withOtherInfoNeededToBuildEvent:(NSDictionary*)whatever.

I would work to dry up code where you init both beacons and events